keylime/ima_ast.py

```diff
@@ -0,0 +1,316 @@
+            self.signature = Signature(tokens[2])
+
+    def hash(self):
+        tohash = self.digest.struct() + self.path.struct()
+        # If no signature is there we sill have to add the entry for it
+        if self.signature is None:
+            tohash += struct.pack("<I0s", 0, b'')
+        else:
+            tohash += self.signature.struct()
+        return hashlib.sha1(tohash).digest()
+
+    def is_data_valid(self, validator):
+        return validator.get_validator(type(self))(self.digest, self.path, self.signature)
+
+
+class ImaBuf(Mode):
+    """
+    Class for "ima-buf". Contains a digest, buffer name and the buffer itself.
+    For validation the buffer must be done based on the name because IMA only provides it as an byte array.
+    """
+    digest: Digest
+    name: Name
+    data: Buffer
+
+    def __init__(self, data: str):
+        tokens = data.split(maxsplit=5)
+        if len(tokens) != 3:
+            raise ParserError(f"Cannot create ImaBuf expected 3 tokens got: {len(tokens)}.")
+        self.digest = Digest(tokens[0])
+        self.name = Name(tokens[1])
+        self.data = Buffer(tokens[2])
+
+    def hash(self):
+        tohash = self.digest.struct() + self.name.struct() + self.data.struct()
+        return hashlib.sha1(tohash).digest()
+
+    def is_data_valid(self, validator: Validator):
+        return validator.get_validator(type(self))(self.digest, self.name, self.data)
+
+
+class Entry:
+    """
+    IMA Entry. Contains the PCR, template hash and mode.
+    """
+    pcr: str
+    template_hash: bytes
+    mode: Mode
+    validator: Validator
+
+    mode_lookup = {
+        "ima": Ima,
+        "ima-ng": ImaNg,
+        "ima-sig": ImaSig,
+        "ima-buf": ImaBuf
+    }
+
+    def __init__(self, data: str, validator=None):
+        self.validator = validator
+        tokens = data.split(maxsplit=3)
+        if len(tokens) != 4:
+            raise ParserError(f"Cannot create Entry expected 4 tokens got: {len(tokens)}.")
+        self.pcr = tokens[0]
+        try:
+            self.template_hash = codecs.decode(tokens[1].encode(), "hex")
+        except binascii.Error as e:
+            raise ParserError(f"Cannot create Entry expected 4 tokens got: {len(tokens)}.") from e
+
+        mode = self.mode_lookup.get(tokens[2], None)
+        if mode is None:
+            raise ParserError(f"No parser for mode {tokens[2]} implemented.")
+        self.mode = mode(tokens[3])
+
+        # TODO lookup if and when this happens
+        if self.template_hash == START_HASH:
+            self.template_hash = FF_HASH
+
+    def valid(self):
+        # TODO why do we need this case?
+        if self.template_hash == FF_HASH:
+            logger.warning("Skipped template_hash validation entry with FF_HASH")
+            return self.mode.is_data_valid(self.validator)
+        if self.template_hash != self.mode.hash():
+            return False
+        if self.validator is None:
+            return False
```

Comment:
Before this gets merged I would like add an explanation why the exception for `START_HASH` and  `FF_HASH` is made and what in IMA causes this.
In an ideal case we could also provide a test IMA log that has this case.

Reply:
We might have to reach out to folks on IRC about this as it seems to be something from the past. It seems it was added in revision 5bcd88dd by @nabilschear in 2017.

Reply:
As there is no locking when opening a file for read and subsequently for write, or the reverse, the file measurement has little value.   IMA defines this as a measurement violation and invalidates the measurement list, by extending the TPM with 0xFF's instead of the 0x00's.  Instead of ignoring these measurement violations, each violation should be reviewed and accounted for.  The most common violations are log files, which should not be included in the measurement list.  Other examples are applications which unnecessarily open files rw.  These applications should be fixed.

If you really want to gloss over these measurement violations, at least don't hard code it.  ima-evm-utils defines the option "--ignore-violations".

Reply:
@mimizohar thank you for your insight. 

@mpeters I'll remove the special handling from this PR and if necessary we can introduce a special option to ignore this violations later.

Reply:
https://wiki.gentoo.org/wiki/Integrity_Measurement_Architecture/Recipes has a nice write up on writing custom policies and excluding log files.

Reply:
And if we remove the special handling now does it change behavior or change behavior for some people?

Reply:
The behavior should only change for people that rely on that those errors are ignored instead of adding those files to the exclude list. The default IMA configuration given in our documentation should only measure executable files if I saw that correctly and this issue should not occur with that configuration.

I can add that feature back with a configuration option which is enabled by default to keep backwards compatibility, but this behavior was not documented/really known before and I would prefer to disable it by default and users can manually enable it if they actually rely on it.

Reply:
I don't think this PR should change the behavior of how keylime works. If we removed handling of the 00- and FF-hashes then not even quote verification would work if those types of entries occurred in the log.

Reply:
You are correct and I've removed the behavior change from this PR.